### PR TITLE
Metrics

### PR DESCRIPTION
**Pull request by jerosalazar**
Este PR agrega un endpoint a la api para obtener la información de tiempo entre creación y asignación de pull requests de un usuario.

### CODE REVIEW

File: spec/commands/get_user_review_request_metrics_spec.rb [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        create(
+          :pull_request_review_request,
+          pull_request: pull_request2,
+          created_at: DateTime.now - 1.year
+        )
+      end
+
+      it "ignores old pull request" do
+        expect(perform(github_user: github_user, limit_month: 9)).to eq(
+          {
```
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:92]: Redundant curly braces around a hash parameter.

File: spec/commands/get_user_review_request_metrics_spec.rb [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+      end
+      let!(:review_request3) do
+        create(
+          :pull_request_review_request,
+          pull_request: pull_request2,
+          created_at: DateTime.now - 1.year
+        )
+      end
+
+      it "ignores old pull request" do
```
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:90]: Example has too many lines [11/10].

File: spec/commands/get_user_review_request_metrics_spec.rb [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+          :pull_request_with_merge,
+          owner: github_user,
+          gh_created_at: DateTime.now - 1.year
+        )
+      end
+      let!(:review_request3) do
+        create(
+          :pull_request_review_request,
+          pull_request: pull_request2,
+          created_at: DateTime.now - 1.year
```
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:86]: Do not use DateTime.now without zone. Use one of Time.zone.now, DateTime.current, DateTime.now.in_time_zone, DateTime.now.utc, DateTime.now.getlocal, DateTime.now.xmlschema, DateTime.now.iso8601, DateTime.now.jisx0301, DateTime.now.rfc3339, DateTime.now.httpdate, DateTime.now.to_i, DateTime.now.to_f instead.

File: spec/commands/get_user_review_request_metrics_spec.rb [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        )
+      end
+    end
+
+    context "with old pull request" do
+      let(:pull_request2) do
+        create(
+          :pull_request_with_merge,
+          owner: github_user,
+          gh_created_at: DateTime.now - 1.year
```
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:79]: Do not use DateTime.now without zone. Use one of Time.zone.now, DateTime.current, DateTime.now.in_time_zone, DateTime.now.utc, DateTime.now.getlocal, DateTime.now.xmlschema, DateTime.now.iso8601, DateTime.now.jisx0301, DateTime.now.rfc3339, DateTime.now.httpdate, DateTime.now.to_i, DateTime.now.to_f instead.

File: spec/commands/get_user_review_request_metrics_spec.rb [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+          :pull_request_review_request,
+          github_user: monkeyci,
+          pull_request: pull_request1,
+          created_at: DateTime.now - 4.minutes
+        )
+      end
+
+      it "ignores review request from monkey" do
+        expect(perform(github_user: github_user, limit_month: 9)).to eq(
+          {
```
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:61]: Redundant curly braces around a hash parameter.

File: spec/commands/get_user_review_request_metrics_spec.rb [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+      let!(:review_request2) do
+        create(
+          :pull_request_review_request,
+          github_user: monkeyci,
+          pull_request: pull_request1,
+          created_at: DateTime.now - 4.minutes
+        )
+      end
+
+      it "ignores review request from monkey" do
```
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:59]: Example has too many lines [11/10].

File: spec/commands/get_user_review_request_metrics_spec.rb [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+      end
+    end
+
+    context "with review request from monkeyci" do
+      let!(:review_request2) do
+        create(
+          :pull_request_review_request,
+          github_user: monkeyci,
+          pull_request: pull_request1,
+          created_at: DateTime.now - 4.minutes
```
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:55]: Do not use DateTime.now without zone. Use one of Time.zone.now, DateTime.current, DateTime.now.in_time_zone, DateTime.now.utc, DateTime.now.getlocal, DateTime.now.xmlschema, DateTime.now.iso8601, DateTime.now.jisx0301, DateTime.now.rfc3339, DateTime.now.httpdate, DateTime.now.to_i, DateTime.now.to_f instead.

File: spec/commands/get_user_review_request_metrics_spec.rb [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        :pull_request_review_request,
+        pull_request: pull_request1,
+        created_at: DateTime.now - 3.minutes
+      )
+    end
+
+    context "with valid date limit" do
+      it "return correct information" do
+        expect(perform(github_user: github_user, limit_month: 9)).to eq(
+          {
```
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:36]: Redundant curly braces around a hash parameter.

File: spec/commands/get_user_review_request_metrics_spec.rb [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    let!(:review_request1) do
+      create(
+        :pull_request_review_request,
+        pull_request: pull_request1,
+        created_at: DateTime.now - 3.minutes
+      )
+    end
+
+    context "with valid date limit" do
+      it "return correct information" do
```
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:34]: Example has too many lines [11/10].

File: spec/commands/get_user_review_request_metrics_spec.rb [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        :pull_request_with_merge,
+        owner: github_user,
+        gh_created_at: DateTime.now - 5.minutes
+      )
+    end
+    let!(:review_request1) do
+      create(
+        :pull_request_review_request,
+        pull_request: pull_request1,
+        created_at: DateTime.now - 3.minutes
```
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:29]: Do not use DateTime.now without zone. Use one of Time.zone.now, DateTime.current, DateTime.now.in_time_zone, DateTime.now.utc, DateTime.now.getlocal, DateTime.now.xmlschema, DateTime.now.iso8601, DateTime.now.jisx0301, DateTime.now.rfc3339, DateTime.now.httpdate, DateTime.now.to_i, DateTime.now.to_f instead.

File: app/commands/get_user_review_request_metrics.rb [hunk 11]
```diff
@@ -0,0 +1,35 @@
+class GetUserReviewRequestMetrics < PowerTypes::Command.new(:github_user, :limit_month)
+  def perform
```
Comment (monkeyci) [inline app/commands/get_user_review_request_metrics.rb:2]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Metrics/AbcSize: Assignment Branch Condition size for perform is too high. [<13, 23, 4> 26.72/25]

Comment (monkeyci) [inline app/commands/get_user_review_request_metrics.rb:2]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Metrics/MethodLength: Method has too many lines. [21/15]

File: app/commands/get_user_review_request_metrics.rb [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    entry_count = 0
+    total_time = 0
+    review_requests_dic = Hash.new
+    limit_date = (Date.today - @limit_month.to_i.months).beginning_of_day
```
Comment (monkeyci) [inline app/commands/get_user_review_request_metrics.rb:6]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Rails/Date: Do not use Date.today without zone. Use Time.zone.today instead.

File: app/commands/get_user_review_request_metrics.rb [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+                                      .where('gh_created_at > ?', limit_date)
+                                      .where.not(merged_by_id: nil)
+    valid_pull_requests.each do |pr|
+      if review_req = pr_has_review_requests(pr)
+        time_delta = (review_req.created_at - pr.gh_created_at).to_i
+        unless time_delta > 1.week
+          entry_count += 1
+          total_time += time_delta
+          review_requests_dic[pr.id] = { pr_title: pr.title, pr_created_at: pr.gh_created_at,
+                                        review_request_created_at: review_req.created_at,
```
Comment (monkeyci) [inline app/commands/get_user_review_request_metrics.rb:17]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Layout/HashAlignment: Align the keys of a hash literal if they span more than one line.

File: app/commands/get_user_review_request_metrics.rb [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+                                        time_delta: time_delta }
```
Comment (monkeyci) [inline app/commands/get_user_review_request_metrics.rb:18]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Layout/HashAlignment: Align the keys of a hash literal if they span more than one line.

File: spec/commands/get_user_review_request_metrics_spec.rb (hunk 9 above)
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:34]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
RSpec/ExampleLength: Example has too many lines [11/10].

File: spec/commands/get_user_review_request_metrics_spec.rb (hunk 6 above)
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:59]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
RSpec/ExampleLength: Example has too many lines [11/10].

File: spec/commands/get_user_review_request_metrics_spec.rb (hunk 2 above)
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:90]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
RSpec/ExampleLength: Example has too many lines [11/10].

File: app/controllers/api/v1/github_users_controller.rb [hunk 15]
```diff
@@ -34,6 +34,15 @@ def organization_recommendation_statistics
     } }, status: :ok
   end
 
+  def user_temporal_metrics
```
Comment (smargozzini) [inline app/controllers/api/v1/github_users_controller.rb:37]: me tinca ponerle un nombre como `pull_requests_information`, porque estamos pidiendo los pull requests y su información. Temporal metrics es muy amplio

File: config/routes.rb [hunk 16]
```diff
@@ -41,6 +41,8 @@
         'github_users#team_score'
       get 'teams/:team_id/users/:github_login/recommendations' =>
         'github_users#team_review_recommendations'
+      get 'users/:github_login/user_temporal_metrics' =>
+        'github_users#user_temporal_metrics'
```
Comment (smargozzini) [inline config/routes.rb:45]: según el comentario de arriba, debería cambiar este también

File: app/commands/get_user_review_request_metrics.rb [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+        end
+      end
+    end
+    review_requests_dic["personal_mean"] = entry_count.positive? ? total_time / entry_count : 0
+    review_requests_dic
+  end
```
Comment (smargozzini) [inline app/commands/get_user_review_request_metrics.rb:24]: me sumo al mono, esta función es un poco larga y costaría entenderla para alguien que le toque usarla en el futuro. trataría de modularizar lo más posible,

File: app/commands/get_user_review_request_metrics.rb [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+
+  def pr_has_review_requests(pull_request)
+    monkey = GithubUser.find_by(login: "monkeyci")
+    review_requests = pull_request.pull_request_review_requests
+                                  .where.not(github_user_id: monkey.id)
+                                  .order(created_at: :asc)
+    if review_requests.count.positive?
+      review_requests[0]
+    end
```
Comment (smargozzini) [inline app/commands/get_user_review_request_metrics.rb:33]: podriai cambiar esto por
`review_requests[0] || []`
y para otro caso, esto `review_requests.count.positive?` es lo mismo que `review_requests.present?`
pero le dejaría como el primero.

File: spec/factories/pull_request_review_requests.rb [hunk 19]
```diff
@@ -1,6 +1,6 @@
 FactoryBot.define do
   factory :pull_request_review_request do
-    association :pull_request
+    association :pull_request, factory: :pull_request
```
Comment (smargozzini) [inline spec/factories/pull_request_review_requests.rb:3]: esta no es necesaria, la agregue yo probando cuando estábamos en la llamada.

File: spec/commands/get_user_review_request_metrics_spec.rb [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+      end
+
+      it "ignores old pull request" do
+        expect(perform(github_user: github_user, limit_month: 9)).to eq(
+          pull_request1.id => {
+            pr_created_at: pull_request1.gh_created_at,
+            pr_title: "Prueba",
+            review_request_created_at: review_request1.created_at,
+            time_delta: 120
+          },
```
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:93]: Trailing whitespace detected.

File: spec/commands/get_user_review_request_metrics_spec.rb [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+  let(:github_user) { create(:github_user) }
+  let!(:monkeyci) { create(:github_user, login: 'monkeyci') }
+
+  def perform(*_args)
+    described_class.for(*_args)
+  end
+
+  context "without pull request" do
+    it "return empty content" do
+      expect(perform(github_user: github_user, limit_month: 9)).to eq({ "personal_mean" => 0 })
```
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:13]: Redundant curly braces around a hash parameter.

File: app/commands/get_user_review_request_metrics.rb [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+                                  .order(created_at: :asc)
+    review_requests[0] || nil
+  end
+
+  def get_valid_pull_requests
+    limit_date = (Time.zone.today - @limit_month.to_i.months).beginning_of_day
+    @github_user.pull_requests.where('gh_created_at > ?', limit_date).where.not(merged_by_id: nil)
+  end
+
+  def get_pr_info(pr, review_req, time_delta)
```
Comment (monkeyci) [inline app/commands/get_user_review_request_metrics.rb:33]: Method parameter must be at least 3 characters long.

Comment (monkeyci) [inline app/commands/get_user_review_request_metrics.rb:33]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Naming/MethodParameterName: Method parameter must be at least 3 characters long.

File: spec/commands/get_user_review_request_metrics_spec.rb (hunk 20 above)
Comment (monkeyci) [inline spec/commands/get_user_review_request_metrics_spec.rb:93]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Layout/TrailingWhitespace: Trailing whitespace detected.

File: app/commands/get_user_review_request_metrics.rb [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+    review_requests_dic["personal_mean"] = entry_count.positive? ? total_time / entry_count : 0
+    review_requests_dic
+  end
+
+  def pr_has_review_requests(pull_request)
+    monkey = GithubUser.find_by(login: "monkeyci")
+    review_requests = pull_request.pull_request_review_requests
+                                  .where.not(github_user_id: monkey.id)
+                                  .order(created_at: :asc)
+    review_requests[0] || nil
```
Comment (smargozzini) [inline app/commands/get_user_review_request_metrics.rb:25]: acá pensé que la idea era retornar una lista vacía si no existía ninguno, en este caso bastaría con `review_requests .first`. Me confundí yo.

File: app/commands/get_user_review_request_metrics.rb [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+    get_valid_pull_requests.each do |pr|
+      if review_req = pr_has_review_requests(pr)
+        time_delta = (review_req.created_at - pr.gh_created_at).to_i
+        unless time_delta > 1.week
+          entry_count += 1
+          total_time += time_delta
+          review_requests_dic[pr.id] = get_pr_info(pr, review_req, time_delta)
+        end
+      end
+    end
```
Comment (smargozzini) [inline app/commands/get_user_review_request_metrics.rb:15]: Esta parte creo que aún se puede hacer mas legible con funciones más chicas.
me imagino algo como 
```
@entry_count = 0 
@total_time = 0 
@review_requests_dic = Hash.new
get_valid_pull_requests.each do |pr|
  add_pr_data_to_hash(pr) # definirla más abajo
end
```
Por lo general cada funcion debe tener un proposito, solo uno. En este caso, esta funcion debe entregar un hash con los datos, entonces los calculos se deben hacer en otr funcion. Si dejamos todo en esta queda difícil de leer y mantener.

File: spec/commands/get_user_review_request_metrics_spec.rb [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+            pr_title: "Prueba",
+            review_request_created_at: review_request1.created_at.to_s,
+            time_delta: 120
+          }, "personal_mean" => 120
+        )
+      end
+    end
+
+    context "with review request from monkeyci" do
+      let!(:review_request2) do
```
Comment (smargozzini) [inline spec/commands/get_user_review_request_metrics_spec.rb:47]: esto es de puro mañoso, pero creo que quedaría más legible si se llama `monkey_review_request` la variable. Y la otra `human_review_request` sin el 1 y el 2.

File: spec/commands/get_user_review_request_metrics_spec.rb [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+            pr_title: "Prueba",
+            review_request_created_at: review_request1.created_at.to_s,
+            time_delta: 120
+          }, "personal_mean" => 120
+        )
+      end
+    end
+
+    context "with old pull request" do
+      let(:pull_request2) do
```
Comment (smargozzini) [inline spec/commands/get_user_review_request_metrics_spec.rb:69]: acá lo mismo, `old_pull_request`. El pull request que está bien lo llamaría `pull_request` nomas.

**Pull request merged**